lib/ansible/modules/clustering/k8s/k8s_auth.py

```diff
@@ -0,0 +1,278 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2018, KubeVirt Team <@kubevirt>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+
+module: k8s_auth
```

Comment:
It may make more sense to call it like `k8s_oauth_token` or something, I would assume I could use this for generic authorization otherwise. Alternatively we could just make it succeed with a no-op in the case that the cluster doesn't support oauth.

Reply:
My thinking in naming it `k8s_auth` was that it should support all the auth mechanisms that will be used for k8s clusters now or in the future. With openshift that's only oauth, but if something somewhere started using zauth or ëauth, that support should be added to this module ideally with some autodetection (which should be trivial).

As to succeeding even when there's no oauth – that'd only be an option if the module first checked for oauth and if that was gone, tried to perform an operation with http basic auth to verify that that in fact works aka that the auth credentials are correct. (Same logic with using client cert instead of user+pwd, should someone add that eventually.) And depending on whether it got oauth or not, the return structure would either contain a token and an empty `password` field or an empty `api_key` field and valid user+pwd. So the fact/anchor creation code from the example would look like this:

```
- set_fact:
    k8s_auth_params: &k8s_auth_params
        host: "{{ k8s_auth_results.k8s_auth.host }}"
        ssl_ca_cert: "{{ k8s_auth_results.k8s_auth.ssl_ca_cert }}"
        verify_ssl: "{{ k8s_auth_results.k8s_auth.verify_ssl }}"
        api_key: "{{ k8s_auth_results.k8s_auth.api_key }}"
        username: "{{ k8s_auth_results.k8s_auth.username }}"
        password: "{{ k8s_auth_results.k8s_auth.password }}"
```

and the `k8s_auth_params` anchor could be used universally as it'd then work both with openshift and k8s clusters. Though again – that would depend on  `api_key` being non–null with openshift and `password` being non–null with http basic auth. I'm not exactly sure that doing the latter (returning a password and storing it) is a valid thing to do in ansible and whether the whole thing isn't too contrived.

@willthames, could you maybe comment as well?

Reply:
makes sense to me, I think in the long term it would be cool to be able to dump in all your parameters and get back a valid param block that you can add to module_defaults.

Reply:
Ok, I've added a comment about this with a link to this particular part of the discussion, so that future changes stay in line with this general design.

As to implementing http basic auth – I've only gotten the go ahead to implement openshift for now (and the description for the module reflects that). If I get a green light from my team to take care of doing the group module defaults thing, I'll make sure to also add http basic auth fallback to this module and have this be as universal and easy to use as possible.

Reply:
By group module_defaults do you mean https://github.com/ansible/ansible/pull/51093 ?

Reply:
Ah, great, having this exist will make my job much easier.